// dmdb/src/driver/queryStream.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, n) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, n) {
            t.__proto__ = n;
        } || function(t, n) {
            for (var e in n) n.hasOwnProperty(e) && (t[e] = n[e]);
        })(t, n);
    };
    return function(t, n) {
        function e() {
            this.constructor = t;
        }
        r(t, n), t.prototype = null === n ? Object.create(n) : (e.prototype = n.prototype, 
        new e());
    };
}(), __awaiter = this && this.__awaiter || function(t, s, o, c) {
    return new (o = o || Promise)(function(e, n) {
        function r(t) {
            try {
                u(c.next(t));
            } catch (t) {
                n(t);
            }
        }
        function i(t) {
            try {
                u(c.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function u(t) {
            var n;
            t.done ? e(t.value) : ((n = t.value) instanceof o ? n : new o(function(t) {
                t(n);
            })).then(r, i);
        }
        u((c = c.apply(t, s || [])).next());
    });
}, __generator = this && this.__generator || function(e, r) {
    var i, u, s, o = {
        label: 0,
        sent: function() {
            if (1 & s[0]) throw s[1];
            return s[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;o; ) try {
                    if (i = 1, u && (s = 2 & n[0] ? u.return : n[0] ? u.throw || ((s = u.return) && s.call(u), 
                    0) : u.next) && !(s = s.call(u, n[1])).done) return s;
                    switch (u = 0, s && (n = [ 2 & n[0], s.value ]), n[0]) {
                      case 0:
                      case 1:
                        s = n;
                        break;

                      case 4:
                        return o.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        o.label++, u = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = o.ops.pop(), o.trys.pop();
                        continue;

                      default:
                        if (!(s = 0 < (s = o.trys).length && s[s.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            o = 0;
                            continue;
                        }
                        if (3 === n[0] && (!s || n[1] > s[0] && n[1] < s[3])) {
                            o.label = n[1];
                            break;
                        }
                        if (6 === n[0] && o.label < s[1]) {
                            o.label = s[1], s = n;
                            break;
                        }
                        if (s && o.label < s[2]) {
                            o.label = s[2], o.ops.push(n);
                            break;
                        }
                        s[2] && o.ops.pop(), o.trys.pop();
                        continue;
                    }
                    n = r.call(e, o);
                } catch (t) {
                    n = [ 6, t ], u = 0;
                } finally {
                    i = s = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var stream_1 = require("stream"), QueryStream = function(e) {
    function t(t) {
        var n = e.call(this, {
            objectMode: !0
        }) || this;
        return n.fetching = !1, t && process.nextTick(function() {
            n._open(t);
        }), n;
    }
    return __extends(t, e), t.prototype._open = function(t) {
        this.rs = t, this.emit("open"), this.emit("metadata", this.rs.metaData);
    }, t.prototype._destroy = function(r, i) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e = this;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return this.rs ? this.fetching ? [ 4, new Promise(function(t) {
                        return e.once("doneFetching", t);
                    }) ] : [ 3, 2 ] : [ 3, 6 ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return t.trys.push([ 2, 4, 5, 6 ]), [ 4, this.rs.do_close() ];

                  case 3:
                    return t.sent(), [ 3, 6 ];

                  case 4:
                    return n = t.sent(), i(n), [ 2 ];

                  case 5:
                    return this.rs = void 0, [ 7 ];

                  case 6:
                    return i(r), [ 2 ];
                }
            });
        });
    }, t.prototype._read = function() {
        return __awaiter(this, void 0, void 0, function() {
            var n, e;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!this.rs) return this.once("open", this._read), [ 2 ];
                    t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 3, 4, 5 ]), this.fetching = !0, this.rs._allowGetRowCall = !0, 
                    [ 4, this.rs.getRow() ];

                  case 2:
                    return n = t.sent(), this.push(n || null), [ 3, 5 ];

                  case 3:
                    return e = t.sent(), this.destroy(e), [ 3, 5 ];

                  case 4:
                    return this.fetching = !1, this.rs ? this.rs._allowGetRowCall = !1 : this.emit("doneFetching"), 
                    [ 7 ];

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, t;
}(stream_1.Readable);

exports.QueryStream = QueryStream;
